### ISSUE TITLE
Cannot handle 401 error during live replication

### ISSUE THREAD

**Issue by amod-joshi** [pouchdb/pouchdb#5607]
Hi,

We have an application running  both on mobile as well on web in Ionic. I am not able to handle the 401 error on mobile when user changes the password on the web. PouchDB keeps on retrying to connect the sync gateway even though sync gateway sends 401 Authentication error. No call back function is called when retry is set true.

If I set the retry to false I can catch 401 error.  But i want retry to set it true since if any other failure conditions it should retry to connect the sync gateway.

**Comment by dharders** [comment]
Currently I get 401 with `localDB.replicate.to/from`, with retry & live true. But it's emitted through the `complete` event (i.e. not the `denied` event, as you would expect).

I'm using CouchDB 2.0 so YMMV as you mentioned you're using sync gateway. Also, I had issues with `sync()`  events not firing, so best to use `replicate()` instead.

P.S. I think it's all about to change with the next release looking at this PR #5421    (which I think should then fire 401's through the `denied` event)

**Comment by amod-joshi** [comment]
Hi Dale,


When are planning your next release so that we can download and check it


Thanks

Amod Joshi


________________________________

**Comment by dharders** [comment]
> When are planning your next release so that we can download and check it

I'm not the maintainer, you'll have to ask these guys  @nolanlawson @daleharvey  ;)

**Comment by nolanlawson** [comment]
> When are planning your next release

This week. :)

**Comment by amod-joshi** [comment]
Hi, I have downloaded the new version of pouchdb(6.0.4)but 401 issues does not seems to be resolved. It is not firing the error in denied event. I am attaching the image for your reference

![errorof pouchdb](https://cloud.githubusercontent.com/assets/21310848/18432781/a02a9632-7901-11e6-8fc2-6ec4baa4d42e.png)

**Comment by dharders** [comment]
Try the 'error' event (I'm getting 403s)

**Comment by amod-joshi** [comment]
I tried that as well but the result is same, 401 error is not getting trap on live sync or on replication

Please find the same code we are using in the application

 syncLive: function () {
        var syncActive = false;
        sync = localDb.sync(remoteDb, {live: true, retry: true})
          .on('active', function (info) {
            console.log('Sync active');
            if (window.cordova) {
              $cordovaToast.showLongBottom('Synchronizing data').then(function (success) {
                // success
                console.log('Sync Toast message displayed');
              }, function (error) {
                // error.
                console.log("Sync toast message error");
              });
            }
            syncActive = true;
          })
          .on('paused', function (info) {
            console.log('Replication up to date or user went offline');
            $rootScope.$broadcast("$db:replicationPaused", syncActive );
            syncActive = false;
          })
          .on('denied', function (err) {
            // a document failed to replicate (e.g. due to permissions)
            console.log('Document failed to replicate (e.g. due to permissions)');
            //sync.cancel();
            console.log(err);
          })
          .on('complete', function (info) {
            console.log('Remote sync complete');
          })
          .on('error', function (err) {
            console.log('Problem encountered during remote database synchronisation');
            console.log(err);
            if (err.status === 401) {
              sync.cancel();
              $ionicHistory.clearCache().then(function () {
                $ionicHistory.clearHistory();
                console.info('Logging out due to sync error');
                $localstorage.setObject('user', {});
                $localstorage.setObject('userData', {});
                $localstorage.setObject('subscriberDetails', {});
                $localstorage.set('deviceToken', "");
                $localstorage.clearLocalStorage();
                $ionicHistory.nextViewOptions({
                  disableAnimate: true,
                  disableBack: true
                });
                $ionicPopup.alert({
                  title: 'Login Failed!',
                  template: 'Your username/password has been changed'
                });
                $state.go('signin');
              });
            }
          });
      },

**Comment by dharders** [comment]
I see your problem. 

Replace sync() with replicate.to() and replicate.from(). 

Ive always found issues with sync() not propagating errors

**Comment by amod-joshi** [comment]
I am able to handle the error in replicate.to() but in replicate.from() 401 is not getting fired at all since i have kept retry true. If i make it false then 401 is getting fired in error event.

I want retry to be true. Please guide me to resolve the issue

Thanks

**Comment by rschramm** [comment]
@dharders @nolanlawson @amod-joshi,

PouchDB is great and we (Amod) are using it to sync with Couchbase in our mobile app.  However, the issue we face with being unable to catch the 401 means that we will need to do a push notification to the app when a password changes on the server so that the user will get logged out of the client and have to reauthenticate with the right credentials, otherwise they blindly go on thinking they are connected and getting updates when they are not.

Is there any timeframe on when this issue will be addressed so we can decide if it makes sense to move forward with this push notification workaround or wait for a (hopefully imminent) fix?

Thanks!

**Comment by daleharvey** [comment]
No timeframes, but if you want any help diagnosing why the 401 is not firing a `denied` event or creating a fix for it we would be happy to help, cheers

**Comment by jsdevel** [comment]
This really is a big issue.  I'm facing it as well.  @daleharvey where is the best place to update this?

**Comment by daleharvey** [comment]
So if you have sync with `retry: true, live:true` then you should be looking for the 

`sync.on("paused", function (err) { 
   if (err) { 
     // the 401 error should be in here
   }
})`

to catch this situation, if you arent seeing it fired then the place to start looking is @ https://github.com/pouchdb/pouchdb/blob/master/packages/node_modules/pouchdb-replication/src/sync.js and https://github.com/pouchdb/pouchdb/blob/master/packages/node_modules/pouchdb-replication/src/replicate.js#L234

I think the best workaround here would be to take `retry:true` off and start using `sync.on("error"`, handling your own retries should be a little easier

**Comment by jsdevel** [comment]
@daleharvey I'm listening on `paused` and don't receive anything with 401.

**Comment by jsdevel** [comment]
So I was able to get `401` with `PouchDB.replicate(src, target, { live: true, retry: true });`; however, it's received in the `change` event:

![change](https://cloud.githubusercontent.com/assets/2605980/19957790/0b919e98-a157-11e6-83d7-9df2af4ba0a2.png)

**Comment by agsv83** [comment]
Hi everyone,

I think I have found part of the issue.

I have tested this against sync gateway 1.2.1 and PouchDB from this build https://github.com/pouchdb/pouchdb/commit/ef1db281fdea6524aa85f1cefa1550efa8d9e3c3

I've seen there is a problem with the error name. On the 'completeReplication' method, when trying to get the non403s variable, the filter uses 'unauthorized'. Using a log I found the error name was 'Unauthorized'. So adding a 'toLowerCase' fixes part of the problem.

As this was very weird I keept looking and found this cames from 'generateErrorFromResponse', where the name of the error is set as 'Unauthorized' from the gateway response. I think this method should check the error code, so for the 401 and others, it could filter using it against allErrors constants to prevent this kind of problems (based on strings comming from 3º parties).

I have also added a "found401s" variable to the "completeReplication" method to force the emission of a denied event and preventing the backoff function in case it is a 401. (this is just related to my problem with this code)

As I don't know very well how to touch this library, your conventions and everything related to event emitting on the pouchdb core, I post it here. By now I will keep using my custom version of the library.

Also, for those who can't fix the browser popup, a solution is found here, http://www.we-edit.de/stackoverflow/question/prevent-authentication-popup-401-with-couchdb-pouchdb-32670580.html, using nginx to change the gateway header. Maybe it is not the best, but I think there is no other way to do so (only recompiling the gateway).

Hope this helps someone else.

Thanks for your support and PouchDB.

**Comment by agsv83** [comment]
Also, I have tested this using two replicators, with replicate.to and replicate.from, instead of using sync.

Sync seams to be broken, it does not propagate the events. At least for me.

**Comment by nolanlawson** [comment]
@agsv83 Sync Gateway is a separate issue, thank you for opening https://github.com/pouchdb/pouchdb/issues/5966

**Comment by nolanlawson** [comment]
Hm no you're right, it's the same issue.

**Comment by agsv83** [comment]
Hi again, 

Only to inform about this, with PouchDB 6.2.0 the only thing required to keep working with this errors from the Sync Gateway is to set the error name to lower case.

On completeReplication, and writeDocs methods, in PouchDB replicate.

I don't know if there are other places that require this change, for me it is working just with that.

**Comment by dharders** [comment]
Thanks @agsv83 for keeping this issue up to date and for your investigation. I just checked it out myself and seems you are correct.

I've submitted a PR (#6888) to hopefully resolve the issue. Just waiting on travis CI and a review now.
